### PR TITLE
Next

### PR DESCRIPTION
**Pull request by sickill**
Next gen player - work in progress 🚧

### CODE REVIEW

File: src/core.js [hunk 1]
```diff
@@ -62,39 +68,29 @@ class AsciinemaPlayerCore {
...
-    let meta;
+    await this.init();
+    const stop = await this.driver.start();
 
-    if (start) {
-      meta = await start;
-    } else {
-      meta = {};
+    if (typeof stop === 'function') {
+      this.driver.stop = stop;
```
Comment (peme14k) [inline src/core.js:82]: n

File: src/index.js [hunk 2]
```diff
@@ -3,6 +3,6 @@ import AsciinemaPlayer from './components/AsciinemaPlayer';
 
 if (window) {
   window.createAsciinemaPlayer = (props, elem) => {
-    render(() => (<AsciinemaPlayer {...props} />), elem);
```
Comment (peme14k) [inline src/index.js]: n

File: src/driver/websocket.js [hunk 3]
```diff
@@ -1,30 +1,17 @@
...
-            rows: rows || data.rows || data.height
-          });
-
-          loaded = true;
-        } else if (data[1] == 'o' && loaded) {
+        if (data[1] == 'o') {
           feed(data[2]);
         }
       }
-
```
Comment (peme14k) [inline src/driver/websocket.js]: P

File: src/index.js (hunk 2 above)
Comment (peme14k) [inline src/index.js]: n

Comment (peme14k) [inline src/index.js]: p

File: src/driver/websocket.js [hunk 4]
```diff
@@ -1,30 +1,17 @@
-function websocket(url, { feed }, { cols, rows }) {
```
Comment (peme14k) [inline src/driver/websocket.js]: n

File: src/driver/websocket.js [hunk 5]
```diff
@@ -1,30 +1,17 @@
...
+function websocket(url, { feed }) {
   let socket;
 
   return {
     start: () => {
-      let resolveLoaded;
```
Comment (peme14k) [inline src/driver/websocket.js]: p

File: src/driver/websocket.js [hunk 6]
```diff
@@ -1,30 +1,17 @@
...
-      let loader = new Promise(resolve => resolveLoaded = resolve);
```
Comment (peme14k) [inline src/driver/websocket.js]: p

File: src/driver/websocket.js [hunk 7]
```diff
@@ -1,30 +1,17 @@
...
-      let loaded = false;
```
Comment (peme14k) [inline src/driver/websocket.js]: n

File: src/driver/test.js [hunk 8]
```diff
@@ -1,26 +1,32 @@
...
-
-    start: () => {
-      intervalId = setInterval(() => { feed(Math.random().toString()) }, 33);
-    },
+  const schedule = () => {
+    const t = Math.pow(5, Math.random() * 4);
+    timeoutId = setTimeout(print, t);
+  }
 
-    stop: () => {
```
Comment (peme14k) [inline src/driver/test.js]: n

File: src/driver/test.js [hunk 9]
```diff
@@ -1,26 +1,32 @@
...
-      clearInterval(intervalId);
```
Comment (peme14k) [inline src/driver/test.js]: p

File: src/driver/asciicast.js [hunk 10]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
       timeoutId = null;
       isFinished = true;
+      pauseElapsedTime = duration * 1000;
       onFinish();
     }
   }
 
   function runFrame() {
     let frame = frames[nextFrameIndex];
-    let actualElapsedTime;
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 11]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
+    let elapsedWallTime;
 
     do {
       feed(frame[1]);
-      virtualElapsedTime += (frame[0] * 1000);
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/driver/asciicast.js [hunk 12]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
+      elapsedVirtualTime += (frame[0] * 1000);
       nextFrameIndex++;
       frame = frames[nextFrameIndex];
-      actualElapsedTime = now() - startedTime;
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 13]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
-    } while (frame && (actualElapsedTime > (virtualElapsedTime + frame[0] * 1000)));
+      elapsedWallTime = now() - startTime;
+    } while (frame && (elapsedWallTime > (elapsedVirtualTime + frame[0] * 1000)));
 
     scheduleNextFrame();
   }
 
-  function start() {
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 14]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
-    nextFrameIndex = 0;
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 15]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
-    virtualElapsedTime = 0;
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 16]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
-    startedTime = now();
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 17]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
-    lastFrameTime = startedTime;
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 18]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
-    isFinished = false;
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js (hunk 12 above)
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 19]
```diff
@@ -1,31 +1,39 @@
...
 // TODO support ttyrec (via opts.format == 'ttyrec')
 
-function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
+function asciicast(url, { feed, now, setTimeout, onFinish }) {
   let frames;
+  let duration;
   let timeoutId;
   let isFinished;
   let nextFrameIndex = 0;
-  let virtualElapsedTime = 0;
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 20]
```diff
@@ -1,31 +1,39 @@
...
-  let startedTime;
```
Comment (peme14k) [inline src/driver/asciicast.js]: n

File: src/driver/asciicast.js [hunk 21]
```diff
@@ -1,31 +1,39 @@
...
-  let lastFrameTime;
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/driver/asciicast.js [hunk 22]
```diff
@@ -1,31 +1,39 @@
...
-  let recording;
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/driver/asciicast.js [hunk 23]
```diff
@@ -1,31 +1,39 @@
...
-
-  function load() {
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/driver/asciicast.js [hunk 24]
```diff
@@ -1,31 +1,39 @@
...
-    if (!recording) {
-      recording = fetch(url).then(res => res.json());
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/driver/asciicast.js [hunk 25]
```diff
@@ -1,31 +1,39 @@
...
-    }
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/driver/asciicast.js [hunk 26]
```diff
@@ -1,31 +1,39 @@
...
+  let meta;
+
+  async function load() {
+    if (!frames) {
+      const res = await fetch(url);
+      const asciicast = await res.json();
+      duration = asciicast['duration'];
+      frames = asciicast['stdout'];
 
-    return recording;
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/driver/asciicast.js [hunk 27]
```diff
@@ -1,31 +1,39 @@
...
+        rows: asciicast['height'],
+        duration: duration
+      };
+    }
   }
 
   function scheduleNextFrame() {
     const nextFrame = frames[nextFrameIndex];
 
     if (nextFrame) {
```
Comment (peme14k) [inline src/driver/asciicast.js:33]: n

File: src/driver/asciicast.js [hunk 28]
```diff
@@ -1,31 +1,39 @@
...
       const delay = nextFrame[0] * 1000;
-      const actualElapsedTime = now() - startedTime;
-      let timeout = (virtualElapsedTime + delay) - actualElapsedTime;
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/driver/asciicast.js [hunk 29]
```diff
@@ -35,55 +43,81 @@ function asciicast(url, { feed, now, setTimeout, onFinish }, { cols, rows }) {
...
+    let elapsedWallTime;
 
     do {
       feed(frame[1]);
-      virtualElapsedTime += (frame[0] * 1000);
+      elapsedVirtualTime += (frame[0] * 1000);
       nextFrameIndex++;
       frame = frames[nextFrameIndex];
-      actualElapsedTime = now() - startedTime;
-    } while (frame && (actualElapsedTime > (virtualElapsedTime + frame[0] * 1000)));
```
Comment (peme14k) [inline src/driver/asciicast.js]: p

File: src/components/AsciinemaPlayer.js [hunk 30]
```diff
@@ -47,15 +47,13 @@ export default props => {
     console.log('mounted!');
 
     if (props.preload) {
-      const preload = core.preload();
+      const meta = await core.preload();
 
-      if (preload) {
-        const { cols, rows } = await preload;
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: n

File: src/components/AsciinemaPlayer.js [hunk 31]
```diff
@@ -109,23 +107,30 @@ export default props => {
   }
 
   const pauseOrResume = () => {
-    const isPlaying = core.pauseOrResume();
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: n

File: src/components/AsciinemaPlayer.js [hunk 32]
```diff
@@ -109,23 +107,30 @@ export default props => {
...
-
-    if (isPlaying) {
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 33]
```diff
@@ -109,23 +107,30 @@ export default props => {
...
-      setState('state', 'playing');
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 34]
```diff
@@ -109,23 +107,30 @@ export default props => {
...
-      startTimeUpdates();
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 35]
```diff
@@ -109,23 +107,30 @@ export default props => {
...
-      startBlinking();
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 36]
```diff
@@ -109,23 +107,30 @@ export default props => {
...
+    if (state.state == 'initial') {
+      play();
     } else {
-      setState('state', 'paused');
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 37]
```diff
@@ -109,23 +107,30 @@ export default props => {
...
-      updateTime();
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 38]
```diff
@@ -109,23 +107,30 @@ export default props => {
...
-      stopTimeUpdates();
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 39]
```diff
@@ -47,15 +47,13 @@ export default props => {
     console.log('mounted!');
 
     if (props.preload) {
-      const preload = core.preload();
+      const meta = await core.preload();
 
-      if (preload) {
-        const { cols, rows } = await preload;
-
-        if (!state.cols) {
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 40]
```diff
@@ -47,15 +47,13 @@ export default props => {
     console.log('mounted!');
 
     if (props.preload) {
-      const preload = core.preload();
+      const meta = await core.preload();
 
-      if (preload) {
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 41]
```diff
@@ -47,15 +47,13 @@ export default props => {
     console.log('mounted!');
 
     if (props.preload) {
-      const preload = core.preload();
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

File: src/components/AsciinemaPlayer.js [hunk 42]
```diff
@@ -47,15 +47,13 @@ export default props => {
...
+      const meta = await core.preload();
 
-      if (preload) {
-        const { cols, rows } = await preload;
-
-        if (!state.cols) {
-          setState({ cols, rows });
```
Comment (peme14k) [inline src/components/AsciinemaPlayer.js]: p

**Review by peme14k — COMMENTED** [review]
Aa E

**Comment by burntcarrot** [comment]
@sickill Apologies for the ping; some questions:

- Why is asciinema-player moving from ClojureScript to `solid-js` (JavaScript)? Since you're proficient with Clojure, I'm curious to know what didn't work out.
- When would the next-gen player be rolled out?
- Would asciinema-server be also ported to some other language in the future? (I see no problem with using Elixir though)

Thanks for creating asciinema, I'm grateful for your contributions! ☺

**Comment by sickill** [comment]
@burntcarrot Hey! No need to apologize! Good questions here.

> Why is asciinema-player moving from ClojureScript to solid-js (JavaScript)? Since you're proficient with Clojure, I'm curious to know what didn't work out.

There are several reasons:

- speed - I wanted the player to be ultra-smooth even for the most heavy animated recordings. Due to ClojureScript's immutable data structures there's a lot of objects created and garbage collected all the time and for the high frame-rate, heavy animations this puts a lot of pressure on CPU and memory. The new implementation of the virtual terminal interpreter in Rust/WASM does it 50x faster (last time I benchmarked, I'll be publishing a blog post about perf actually)
- size - the output bundle from ClojureScript compiler is rather relatively big. It's fine when you build your own app in ClojureScript but when you provide a library to use by other people in their websites it's rather terrible (almost 600kb minified). The new version uses Solid.js instead of React.js which makes it smaller, the fact that we no longer ship whole ClojureScript standard library in the bundle helps making it even smaller. The new player is now 137kb, even though it includes embeded WASM bytecode (base64 encoded).
- integration with JS ecosystem - ClojureScript is not that easy to integrate with the JS ecosystem. I know there's been a lot of improvements done in this space over the years but it's still extra work you need to do compared to regular JS codebase, and some things didn't have any support last time I checked (like embedding WASM in the bundle). Things might have changed here but above first two arguments still apply so it was worth it :)

And I had fun re-writing the terminal interpreter in Rust :)

> When would the next-gen player be rolled 

It's around the corner. I just merged the `next` branch into `develop`, published beta version of the package to npm and updated README. There's one or two small things I'd like to do but in general it can be used already. Feel free to send a feedback if you happen to try it out.

> Would asciinema-server be also ported to some other language in the future? (I see no problem with using Elixir though)

No. There's no need for that. Elixir and Phoenix are excellent and they work just perfectly fine for the server part. They have good things like LiveView for example which could be used to make the site nicer.

> Thanks for creating asciinema, I'm grateful for your contributions! ☺

Thanks for the nice words 🙌 ❤️

**Comment by qnkhuat** [comment]
Hi @sickill, 
Does the next version add support for resizing?

Also, looking forward for your blog on switching to JS :)

**Comment by sickill** [comment]
@qnkhuat yes! Automatic resizing of the player to fill the available space is in place now.

**Comment by sickill** [comment]
@qnkhuat here's the blog https://blog.asciinema.org/post/smaller-faster/

**Comment by qnkhuat** [comment]
Thanks for sharing @sickill !

**Pull request merged**